Fix sparse BSR tensor validation.

pearu opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/78264

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78173
* __->__ #78359

Also adds bits to support dense dimensions for Sparse Compressed tensors.

pearu committed:
Fix sparse BSR tensor validation.

Also adds bits to support dense dimensions for Sparse Compressed tensors.

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -103,38 +103,49 @@ void _validate_sparse_compressed_tensor_args_worker(const Tensor& compressed_ind
       compressed_indices.dim(), " and ", plain_indices.dim(), ", respectively");
 
-  AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(
+  int block_ndim = AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(
```
For follow up: I'm wondering whether we should store this as a member variable of SparseCsrTensorImpl.h. Really its value can only be 0 or 2.

pearu replied:
Currently, the `blocksize` can be uniquely determined from the sparse tensor `layout` and the shape of `values` tensor (`blocksize` is `values().shape[1:3]` for BSR/BSC tensors). That said, we could implement a `blocksize()` method for `SparseCsrTensorImpl` that DimVector instance of a blocksize according to the layout and values info.

pearu replied:
I suggest that we address this when implementing hybrid sparse compressed tensors support.

pearu replied:
Although, when calling `validate_sparse_compressed_tensor_args` function, our inputs are the arguments to sparse compressed tensor factory functions. So, introducing `blocksize()` method for SparseCsrTensorImpl is out of scope of the function.

cpuhrsch reviewed (commented):
Great! Thanks for sending this!

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -103,38 +103,49 @@ void _validate_sparse_compressed_tensor_args_worker(const Tensor& compressed_ind
+      " (minus the number of dense dimensions) but got ",
+      compressed_indices.dim(), " not equal to ", size.size(), " - 1 - ", dense_ndim);
+
+  DimVector blocksize{
```
Part of me does wonder whether it's more "sanitary" so to speak to separate out this validation into two separate functions. Essentially what this shows, and which is true, is that we can think of a regular CSR Tensor as a BSR Tensor with 1 by 1 blocks. Maybe eventually SparseTensorImpl.h will be able to do it all.

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -363,23 +365,29 @@ def test_print(self, layout, device):
 
         def run_test(shape, nnz, index_type):
-            block_size = (2, 3) if layout in {torch.sparse_bsr, torch.sparse_bsc} else ()
+            blocksize = tuple(map(self._smallest_divisor, shape[-2:])) if layout in {torch.sparse_bsr, torch.sparse_bsc} else ()
```
Why not hardcode the values instead of writing this function? I agree that it's cleaner, but for tests transparency matters most (to me at least).

pearu replied:
It would just mean that `run_test` function would need to be extended with `blocksize` argument. Makes sense to me.

pearu replied:
Fixed.

cpuhrsch commented:
You'll need to merge master into this to fix the CI errors.

pearu committed:
Update on "Fix sparse BSR tensor validation."

Fixes https://github.com/pytorch/pytorch/issues/78264


Also adds bits to support dense dimensions for Sparse Compressed tensors.

[ghstack-poisoned]

pearu committed:
Update on "Fix sparse BSR tensor validation."

Fixes https://github.com/pytorch/pytorch/issues/78264


Also adds bits to support dense dimensions for Sparse Compressed tensors.

[ghstack-poisoned]

pearu commented:
@pytorchbot merge on green
